features/atcryptoauth/sample_ssl_client.cpp

```diff
@@ -0,0 +1,393 @@
+"VExTMQswCQYDVQQGEwJVSzBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABOC03rQc\r\n"
+"nePEsBd8IVKaBStz2LfRNAXnxbyR7vbdP2qvF72KHfcwj3/EieTz7o0v399J9xBe\r\n"
+"Gjviq0DPMLRYpq6jLDAqMAkGA1UdEwQCMAAwHQYDVR0OBBYEFG973jHQ2GP8JXd2\r\n"
+"9HUvNT2GWZV2MAwGCCqGSM49BAMCBQADSAAwRQIgP9Dv55CBaj2ZE+XVxkUizE0A\r\n"
+"GsdZvOiFnlOICCOyB4gCIQCFqRLUkctB3Anxfqz6E3hjlGk68i5PNnXDvcRkVzQE\r\n"
+"Wg==\r\n"
+"-----END CERTIFICATE-----\r\n";
+
+static const char ca_cert_pem[] =
+"-----BEGIN CERTIFICATE-----\r\n"
+"MIICfDCCAiGgAwIBAgIBADAKBggqhkjOPQQDAjCBnTELMAkGA1UEBhMCR0IxFzAV\r\n"
+"BgNVBAgMDkNhbWJyaWRnZXNoaXJlMRIwEAYDVQQHDAlDYW1icmlkZ2UxDjAMBgNV\r\n"
+"BAoMBU15IENvMRAwDgYDVQQLDAdNeSBEZXB0MR0wGwYDVQQDDBRtYXppbWtoYW4t\r\n"
+"VmlydHVhbEJveDEgMB4GCSqGSIb3DQEJARYRQXppbS5LaGFuQGFybS5jb20wHhcN\r\n"
+"MTgwMjAzMDEzMzA3WhcNMjgwMjAxMDEzMzA3WjCBnTELMAkGA1UEBhMCR0IxFzAV\r\n"
+"BgNVBAgMDkNhbWJyaWRnZXNoaXJlMRIwEAYDVQQHDAlDYW1icmlkZ2UxDjAMBgNV\r\n"
+"BAoMBU15IENvMRAwDgYDVQQLDAdNeSBEZXB0MR0wGwYDVQQDDBRtYXppbWtoYW4t\r\n"
+"VmlydHVhbEJveDEgMB4GCSqGSIb3DQEJARYRQXppbS5LaGFuQGFybS5jb20wWTAT\r\n"
+"BgcqhkjOPQIBBggqhkjOPQMBBwNCAAS+ImO4q+2w6l8fN3y9dRFgmb4dmaCYIUK6\r\n"
+"ZG0hGT/AqLC8ds0al7Mzw8+MUsYNpsxUToFU4EvV3iCxhd2vZFhho1AwTjAdBgNV\r\n"
+"HQ4EFgQUBN3Eifs+3C9ZwHMAF3GBPNKDH08wHwYDVR0jBBgwFoAUBN3Eifs+3C9Z\r\n"
+"wHMAF3GBPNKDH08wDAYDVR0TBAUwAwEB/zAKBggqhkjOPQQDAgNJADBGAiEAhYwb\r\n"
+"wAQiGkOTt/YbdPY1ZuEjCBY6dyqCPI9blT0EVvQCIQCJCSLSQ0IW37SI4RHXuBDZ\r\n"
+"fCh27v3WWb8djrwpjSyg+g==\r\n"
+"-----END CERTIFICATE-----\r\n";
+
+static TCPSocket * _tcpsocket = NULL;
+
+int mbedtls_net_init()
+{
+    EthernetInterface eth_iface;
+
+    mbedtls_printf( "\r\n************************* GET IP ADDRESS *************************\r\n" );
+    eth_iface.connect();
+    const char *ip_addr = eth_iface.get_ip_address();
+    if (ip_addr)
+        mbedtls_printf("Client IP Address is %s\r\n", ip_addr);
+    else
+    {
+        mbedtls_printf("No Client IP Address\r\n");
+        return 1;
+    }
+
+    _tcpsocket = new TCPSocket(&eth_iface);
+    _tcpsocket->set_blocking(false);
+    return 0;
+}
+
+int mbedtls_net_connect(const char * server_name, uint16_t port)
+{
+    /* Connect to the server */
+    mbedtls_printf("Connecting with %s\n", server_name);
+    int ret = _tcpsocket->connect(server_name, port);
+    if (ret != NSAPI_ERROR_OK) {
+        mbedtls_printf("Failed to connect\n");
+        printf("MBED: Socket Error: %d\n", ret);
+        _tcpsocket->close();
+        return 1;
+    }
+    return 0;
+}
+
+void mbedtls_net_free()
+{
+    _tcpsocket->close();
+    delete _tcpsocket;
+}
+
+static int mbedtls_net_send(void *ctx, const unsigned char *buf, size_t len) {
+    int size = -1;
+    TCPSocket *socket = static_cast<TCPSocket *>(ctx);
+    size = socket->send(buf, len);
+
+    if(NSAPI_ERROR_WOULD_BLOCK == size){
+        return MBEDTLS_ERR_SSL_WANT_WRITE;
+    }else if(size < 0){
+        mbedtls_printf("Socket send error %d\n", size);
+        return -1;
```

Comment:
This should be an `MBEDTLS_ERR_SSL_XXX` error code, translated from the error returned by `TCPSocket::send`.

Same remark in `mbedtls_net_recv`.